src/Storages/StorageURL.cpp

```diff
@@ -735,6 +736,11 @@ Pipe IStorageURLBase::read(
     {
         /// Iterate through disclosed globs and make a source for each file
         auto glob_iterator = std::make_shared<StorageURLSource::DisclosedGlobIterator>(uri, max_addresses, query_info.query, virtual_columns, local_context);
+
+        /// check if we filtered out all the paths
+        if (glob_iterator->size() == 0)
+            return Pipe(std::make_shared<NullSource>(storage_snapshot->getSampleBlockForColumns(column_names)));
```

Comment:
Hmm, just noticed. `column_names` may contain virtual columns that are not in storage_snapshot, so `getSampleBlockForColumns` can throw an exception. We can init `read_from_format_info` before iterators and use `read_from_format_info.source_header`